src/main/java/rx/observers/TestSubscriber.java

```diff
@@ -271,17 +318,74 @@ public void requestMore(long n) {
      *          if the sequence of items observed does not exactly match {@code items}
      */
     public void assertReceivedOnNext(List<T> items) {
-        testObserver.assertReceivedOnNext(items);
+        if (values.size() != items.size()) {
+            assertionError("Number of items does not match. Provided: " + items.size() + "  Actual: " + values.size()
+            + ".\n"
+            + "Provided values: " + items
+            + "\n"
+            + "Actual values: " + values
+            + "\n");
+        }
+
+        for (int i = 0; i < items.size(); i++) {
+            T expected = items.get(i);
+            T actual = values.get(i);
+            if (expected == null) {
+                // check for null equality
+                if (actual != null) {
+                    assertionError("Value at index: " + i + " expected to be [null] but was: [" + actual + "]\n");
+                }
+            } else if (!expected.equals(actual)) {
+                assertionError("Value at index: " + i 
+                        + " expected to be [" + expected + "] (" + expected.getClass().getSimpleName() 
+                        + ") but was: [" + actual + "] (" + (actual != null ? actual.getClass().getSimpleName() : "null") + ")\n");
+
+            }
+        }
     }
 
     /**
+     * Wait until the current committed value count is less than the expected amount
+     * by sleeping 1 unit at most timeout times and return true if at least
+     * the required amount of onNext values have been received.
+     * @param expected the expected number of onNext events
+     * @param timeout the time to wait for the events
+     * @param unit the time unit of waiting
+     * @return true if the expected number of onNext events happened
+     * @throws InterruptedException if the sleep is interrupted
+     * @since (if this graduates from Experimental/Beta to supported, replace this parenthetical with the release number)
+     */
+    @Experimental
+    public final boolean awaitValueCount(int expected, long timeout, TimeUnit unit) throws InterruptedException {
+        while (timeout != 0 && valueCount < expected) {
+            unit.sleep(1);
```

Comment:
@JakeWharton well, it depends on use case, I guess somebody who wants to use `await(time)` on `TestSubscriber` either wants it to be:

* "max timeout to fail the test instead of looping infinitely", like @Timeout in JUnit. so in that case value will be someting like `1, MINUTE` while actually data will arrive much faster.
* "precise value to check some concurrency algorithm with expected timeouts" so in that case "success" long (relatively) after actual timeout can be considered as a bug.

I'd be ok with something like this:

```java
while (true) {
  if (valueCount >= expected) {
    return true;
  }

  if (System.nanoTime() - start > timeoutNano) {
    return false;
  }

  Thread.sleep(1);
}
```

So that only flakiness of last `sleep()` will be amortized.

Current implementation increments `timeout` after each `sleep(1)` and may collect some relatively big error, like you had 150 `sleep(1)` but actually spent `200ms` which is ~25% error and seems possible in real life.